lenskit-eval/src/main/java/org/lenskit/eval/traintest/recommend/TopNNDPMMetric.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.eval.traintest.recommend;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import it.unimi.dsi.fastutil.longs.*;
+import org.apache.commons.lang3.StringUtils;
+import org.grouplens.lenskit.util.statistics.MeanAccumulator;
+import org.lenskit.api.ResultList;
+import org.lenskit.eval.traintest.AlgorithmInstance;
+import org.lenskit.eval.traintest.DataSet;
+import org.lenskit.eval.traintest.TestUser;
+import org.lenskit.eval.traintest.metrics.MetricResult;
+import org.lenskit.specs.AbstractSpec;
+import org.lenskit.util.collections.LongUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+/**
+ * Measure the nDPM of the top-N recommendations, using rankings.
+ * This metric is registered with the type name `ndpm`.
+ * The paper used as a reference for this implementation is http://www2.cs.uregina.ca/~yyao/PAPERS/jasis_ndpm.pdf.
+ * Created by VaibhavMahant on 2/1/16.
```

Comment:
We do not generally sign our code files — the relevant information is in the Git committer log.